### PR TITLE
Text Tower Support

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

**Comment by iejMac** [comment]
@rom1504  @rwightman  This PR just moves the TextTower stuff out of the GradCache PR and updates the components to what they are in main now.

Training on 1 GPU works fine.

File: src/open_clip/factory.py [hunk 1]
```diff
@@ -56,6 +56,8 @@ def load_state_dict(checkpoint_path: str, map_location='cpu'):
         state_dict = checkpoint
     if next(iter(state_dict.items()))[0].startswith('module'):
         state_dict = {k[7:]: v for k, v in state_dict.items()}
+    if 'positional_embedding' in state_dict:
```
Comment (rom1504) [inline src/open_clip/factory.py:59]: Why is this needed

File: src/open_clip/model.py [hunk 2]
```diff
@@ -543,35 +124,17 @@ def set_grad_checkpointing(self, enable=True):
...
-        x = self.token_embedding(text)  # [batch_size, n_ctx, d_model]
-
-        x = x + self.positional_embedding
-        x = x.permute(1, 0, 2)  # NLD -> LND
-        x = self.transformer(x, attn_mask=self.attn_mask)
-        x = x.permute(1, 0, 2)  # LND -> NLD
-        x = self.ln_final(x)
+    def encode_image(self, image, normalize: bool = False):
+        features = self.visual(image)
+        return F.normalize(features, dim=-1) if normalize else features
```
Comment (rom1504) [inline src/open_clip/model.py:129]: This is a breaking change as before the normalize was done outside

File: src/open_clip/model.py [hunk 3]
```diff
@@ -599,6 +162,24 @@ def _convert_weights_to_fp16(l):
     model.apply(_convert_weights_to_fp16)
 
 
+def convert_state_dict(state_dict: dict):
+    if 'text_projection' in state_dict:
```
Comment (rom1504) [inline src/open_clip/model.py:166]: Is this used to maintain checkpoint compatibility?
Please add a comment

File: src/open_clip/modified_resnet.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
+from collections import OrderedDict
```
Comment (rom1504) [inline src/open_clip/modified_resnet.py:1]: Is there any diff compared to code in main ?

File: src/open_clip/transformer.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
+from collections import OrderedDict
```
Comment (rom1504) [inline src/open_clip/transformer.py:1]: Is there any diff compared to code in main ?

File: src/open_clip/model.py (hunk 2 above)
Comment (rom1504) [reply src/open_clip/model.py:129]: Let's undo this part (the normalize)

File: src/open_clip/factory.py (hunk 1 above)
Comment (rom1504) [reply src/open_clip/factory.py:59]: Ok i see, checkpoint compatibility
Let's add a comment

Comment (iejMac) [reply src/open_clip/factory.py:59]: text tower change will make state dicts from before change and after change incompatible so convert_state_dict fixes that and I think
```
'positional_embeddings' in state_dict
```
just detects that. Please correct me if I'm wrong @rwightman

File: src/open_clip/model.py [hunk 6]
```diff
@@ -2,414 +2,21 @@
...
 
 import numpy as np
 import torch
 import torch.nn.functional as F
 from torch import nn
 from torch.utils.checkpoint import checkpoint
 
+from .modified_resnet import ModifiedResNet
 from .timm_model import TimmModel
-from .utils import freeze_batch_norm_2d, to_2tuple
```
Comment (rom1504) [inline src/open_clip/model.py:18]: Is there any diff with the moved code ?

File: src/open_clip/model.py (hunk 2 above)
Comment (iejMac) [reply src/open_clip/model.py:129]: yeah, it's also pretty nice. you think we should remove it?

Comment (rom1504) [reply src/open_clip/model.py:129]: Actually ok because false by default

File: src/open_clip/modified_resnet.py (hunk 4 above)
Comment (iejMac) [reply src/open_clip/modified_resnet.py:1]: updated now so all is copied from main

File: src/open_clip/transformer.py (hunk 5 above)
Comment (iejMac) [reply src/open_clip/transformer.py:1]: no

File: src/open_clip/factory.py (hunk 1 above)
Comment (rom1504) [inline src/open_clip/factory.py:59]: Would be good to add the comment saying it's for compatibility here too

**Comment by rwightman** [comment]
@iejMac this is looking good thanks, will definitely need some extensive testing before a final merge, the usage of this repo has gone up quite a bit recently

Also, further that point, in the original design I changed the one and only CLIP model class to have a separate tower, added the convert fn figuring there weren't that many users to confuse until they figured that out. With broader use now,  that's probably not the best approach. I feel we should move have two model classes now, one as is that doesn't change, and and a new one that's used for configs that enable custom text towers. We can still provide conversion fns to make it easy for one to take an existing non TT model and adapt it to a TT config for fine-tune or further train.

Thoughts? Re naming not quite sure .... ideas
* `CLIPT`
* `CustomTextCLIP`
* ...?

@rom1504 any concerns with this approach?

**Comment by rom1504** [comment]
What are drawbacks to the way this PR is currently done ?

Ie why do we need 2 classes vs just changing the existing one (while keeping compatibility) ?

**Comment by rwightman** [comment]
I don't see a way to keep code clean and have compatibility without forcing users to covert weights, conversion can be automatic for pretained factory use, but not if the models are being integrated non trivially into larger models, etc

**Comment by rom1504** [comment]
The purpose of this PR is to make text tower support core to the lib.

That said I understand your point, probably best to keep perfect checkpoint compatibility.

About the name of the new one, could be CustomTextCLIP i guess.

**Comment by rwightman** [comment]
@rom1504 by 'core feature' I meant most among the most important features. It may get there someday, or may remain a niche use case. At this point I don't think it's worth frustrating the users that have been integrating the recent large models into other systems. But do very much think it's worth adding as there are people who want to experiment with it.

Say if we trained some more checkpoints right after this merge (that don't use this feature), they would have the new state dict format, they'd be different than existing ones, lots of downstream use may require extra code / fixes, remaping old checkpoints isn't likely, having different state dict formats for models that look like they should be part of the same series/family seems off and will confuse people.

Users who've started doing things like

``` python
class MyModel(nn.Module)
  def __init__():
    self.clip = CLIP()
    self.other = Other()
```
and then use standard torch load/save state dict will be upset... etc



The custom text tower models would already stand out as different so I feel it's not an issue for them to use a different model class, the code of the model itself is fairly minimal.

**Comment by rom1504** [comment]
Can you rebase on main head ? I want to see the (new) test fail

**Comment by rom1504** [comment]
Looks like we need a test with direct use of the Clip class since we want to maintain compatibility to that (and it would fail here)
